### ISSUE TITLE
Allow parallel execution of EvalExpression

### ISSUE THREAD

**Issue by andrew-vandenbrink** [fsharp/FSharp.Compiler.Service#482]
I believe we have serious bottle neck (and maybe locking) with the way we call f# evaluation session
This code is based on the FSharp.Compiler.Service documentation
```
public static FsiEvaluationSession Create(string fsiExePath)
        {
            string[] arguments = new string[2] { fsiExePath, "--noninteractive" };

            var sbOut = new StringBuilder();
            var sbErr = new StringBuilder();
            var inStream = new StringReader("");
            var outStream = new StringWriter(sbOut);
            var errStream = new StringWriter(sbErr);

            var config = FsiEvaluationSession.GetDefaultConfiguration();
            var option = new FSharpOption<bool>(true);
            
            return FsiEvaluationSession.Create(config, arguments, inStream, outStream, errStream, option);
        }

        public static T Eval<T>(string fsiExePath, string script)
        {
            var session = Create(fsiExePath);            
            var option = session.EvalExpression(script);
            return (T)option.Value.ReflectionValue;
        }

//Code to change
var eval = await Task.Run(() => FSharpAPI.Eval<bool>(FileAPI.RelativeExecutablePath(@"F#\fsi.exe"), $"if {condition} then true else false"));
```
![2015-12-03 14_39_04-clipboard f](https://cloud.githubusercontent.com/assets/12871379/11564156/0001f8b6-99d8-11e5-9776-1a7564cdab63.png)

Performance take 1 minute

Changing the exact code line to c# version
```
public static async Task<T> EvalAsync<T>(string script)
        {
            return await CSharpScript.EvaluateAsync<T>(script);          
        }
//Changed code line below
var eval = await CSharpAPI.EvalAsync<bool>($"if ({condition}) {{ return true; }} else {{ return false; }}");
```
![2015-12-03 16_03_55-clipboard c](https://cloud.githubusercontent.com/assets/12871379/11564188/346cd698-99d8-11e5-994a-d548ba74c7e5.png)

Performance is 4 seconds now...

My use case is that I need to create a lot of independent of script evaluation. Imagine you have 10 independent scripts then you evaluate it in parallell. 

Do I do it wrong in f# version? is there any other way to evaluate f# script? if not then I suggest that we have f# version of CSharpScript.EvaluateAsync<T>(script)

On f# version I can feel that fsi is somehow locked, where lot of threads can't access it.

**Comment by dsyme** [comment]
There's something very strange. This should take at most 6 seconds (first time, to JIT the compiler) and less on subsequent invocations.  If it always took 1 minute our unit tests hardly ever terminate :)

**Comment by dsyme** [comment]
Is the code in {condition} that take 1 minute to complete?

Adding async execution is possible I suppose. At the moment the actual code is executed in the calling thread.  But using Task.Run should I think make it effectively async via the thread pool

**Comment by andrew-vandenbrink** [comment]
This is the same picture with the first picture but with show external code.
![2015-12-04 07_27_09-clipboard](https://cloud.githubusercontent.com/assets/12871379/11583322/cb773be8-9a58-11e5-8004-7478dd5c4b13.png)


{Condition} take seconds, but I call it more than once
f# version
![2015-12-04 07_35_08-clipboard](https://cloud.githubusercontent.com/assets/12871379/11583487/b28bda6a-9a5a-11e5-8926-66a21c1d565d.png)

c# version
![c eval](https://cloud.githubusercontent.com/assets/12871379/11583496/d76291b2-9a5a-11e5-9a46-2776b56c6262.JPG)

Condition value is simple as this
![condition](https://cloud.githubusercontent.com/assets/12871379/11583502/edc9f710-9a5a-11e5-8507-474aeb5cf2f5.JPG)


Do you have any stress test within the unit test? Something like this
```
for(int i=0;i<1000;i++)
{
Console.WriteLine(i);
var x=FSharpAPI.Eval<int>("path to fsi.exe","1+1");
var y=FSharpAPI.Eval<int>("path to fsi.exe","2+1");
var z=FSharpAPI.Eval<int>("path to fsi.exe","3+1");

Console.WriteLine(x);
Console.WriteLine(y);
Console.WriteLine(z);
}
```
Very slow

**Comment by dsyme** [comment]
Could you change the tile to reflect what issue you're reporting?  Is it

1. that FSharp.Compiler.Service compilation is slower than C# Roslyn compilation (when used synchronously)

2. that the code generated by the F# compiler is slower than the code generated by the C# compiler (when evaluated synchronously)

3. that the F# compiler doesn't provide an asynchronous compile+evaluate primitive (which is not magically faster, it just pushes compilation and execution off to a background thread or queue)

Currently the title looks like (3).

If you're reporting issue 1, then try not to recreate the entire FsiEvaluationSession every time you invoke. Call this once:

           var session = Create(fsiExePath);            

then call this multiple times:

            var option = session.EvalExpression(script);

**Comment by andrew-vandenbrink** [comment]
```
private static FsiEvaluationSession session = null;
        public static FsiEvaluationSession Create(string fsiExePath)
        {
            if (session == null)
            {
                string[] arguments = new string[2] { fsiExePath, "--noninteractive" };

                var sbOut = new StringBuilder();
                var sbErr = new StringBuilder();
                var inStream = new StringReader("");
                var outStream = new StringWriter(sbOut);
                var errStream = new StringWriter(sbErr);

                var config = FsiEvaluationSession.GetDefaultConfiguration();
                var option = new FSharpOption<bool>(true);

                session = FsiEvaluationSession.Create(config, arguments, inStream, outStream, errStream, option);
            }
            return session;
        }
```
i have to use private static variable (and lock), not very functional :P 

I believe this "cache" mechanism should be hidden inside the FSharp.Compiler.Service API's FsiEvaluationSession.Create

I will change the title to "Implement automatic reuse of session which is created by FsiEvaluationSession.Create"
I think thats what Roslyn c# did also but it is hidden to the user of the API

do you agree?

**Comment by andrew-vandenbrink** [comment]
Sorry, I get new issue now
```
 var option = session.EvalExpression(script);
```

As I reuse session, when I call session.EvalExpression in parallel
I get evaluation failed randomly

I assume EvalExpression always have independent context (and thread safe)

![duplicate type name](https://cloud.githubusercontent.com/assets/12871379/11587472/fadddc4a-9a79-11e5-9560-ac72ad5425c0.JPG)

![duplicate type name session](https://cloud.githubusercontent.com/assets/12871379/11587474/04971c6a-9a7a-11e5-8371-a14d3a954bde.JPG)


Conclusion at this moment:
Reuse session -> can't call it in parallel
Always create session -> very slow

**Comment by dsyme** [comment]
@andrew-vandenbrink To execute in parallel, please submit async or task-valued expressions, e.g.

    let evalExpressionTyped<'T> (text) = 
        match fsiSession.EvalExpression(text) with
        | Some value -> value.ReflectionValue |> unbox<'T>
        | None -> failwith "Got no result!"

    open System.Threading.Tasks

    let sampleAsyncExpr = 
        """
    async { do System.Threading.Thread.Sleep 5000
            return 10 }
      |> Async.StartAsTask"""

    let task1 = evalExpressionTyped<Task<int>>(sampleAsyncExpr)  
    let task2 = evalExpressionTyped<Task<int>>(sampleAsyncExpr)  

    task1.Result
    task2.Result


The body of the ``async { ... }`` can do anything you like.

**Comment by dsyme** [comment]
The documentation for this technique has been updated here: https://github.com/fsharp/FSharp.Compiler.Service/blob/7e7bcdd9fed2bb8ee94d09c21b4bfa7ff3d35a6c/docs/content/interactive.fsx#L170

**Comment by dsyme** [comment]
Closing this as evaluating a task should be sufficient
